stackslib/src/net/stackerdb/config.rs

```diff
@@ -0,0 +1,457 @@
+        (
+            "stackerdb-get-signer-slots".into(),
+            TypeSignature::new_response(
+                ListTypeData::new_list(
+                    TupleTypeSignature::try_from(vec![
+                        ("signer".into(), TypeSignature::PrincipalType),
+                        ("num-slots".into(), TypeSignature::UIntType)
+                    ])
+                    .expect("FATAL: failed to construct signer list type")
+                    .into(),
+                    MAX_HINT_REPLICAS
+                )
+                .expect("FATAL: could not construct signer list type")
+                .into(),
+                TypeSignature::UIntType
+            ).expect("FATAL: failed to construct response with signer slots"),
+        ),
+        (
+            "stackerdb-get-config".into(),
+            TypeSignature::new_response(
+                TypeSignature::TupleType(
+                    TupleTypeSignature::try_from(vec![
+                        ("chunk-size".into(), TypeSignature::UIntType),
+                        ("write-freq".into(), TypeSignature::UIntType),
+                        ("max-writes".into(), TypeSignature::UIntType),
+                        ("max-neighbors".into(), TypeSignature::UIntType),
+                        ("hint-replicas".into(), ListTypeData::new_list(
+                            TypeSignature::TupleType(
+                                TupleTypeSignature::try_from(vec![
+                                    ("addr".into(), ListTypeData::new_list(TypeSignature::UIntType, 16)
+                                        .expect("FATAL: invalid IP address list")
+                                        .into()),
+                                    ("port".into(), TypeSignature::UIntType),
+                                    ("public-key-hash".into(),
+                                        // can't use BUFF_20 here because it's also in a
+                                        // lazy_static! block
+                                        TypeSignature::SequenceType(SequenceSubtype::BufferType(BufferLength::try_from(20u32).expect("FATAL: could not create (buff 20)"))))
+                                ])
+                                .expect("FATAL: unable to construct hint-replicas type")
+                                .into()),
+                            MAX_HINT_REPLICAS)
+                            .expect("FATAL: failed to construct hint-replicas list type")
+                            .into())
+                    ]).expect("FATAL: unable to construct config type")).into(),
+                TypeSignature::UIntType
+            ).expect("FATAL: unable to construct config response type")
+        )
+    ];
+}
+
+impl StackerDBConfig {
+    /// Check that a smart contract is consistent with being a StackerDB controller
+    fn is_contract_valid(epoch: &StacksEpochId, analysis: ContractAnalysis) -> bool {
+        for (name, func_return_type) in REQUIRED_FUNCTIONS.iter() {
```

Comment:
This could be simplified if we required stackerdb controller contracts to explicitly use `impl-trait` then, just check `analysis.implemented_traits`. It should be a good practice for those contracts to do that anyway, since it forces a check on implementing the trait at deployment time.

Reply:
I did consider this, and would like to do it, but that would (1) preclude using contracts that do not explicitly `impl-trait` (we'd have to check the contract against the trait), and (2) would preclude using contracts whose stackerdb-related functions are read-only.

Reply:
Will track this in an issue: https://github.com/stacks-network/stacks-blockchain/issues/3890